Source/Core/Core/IOS/DolphinDevice.cpp

```diff
@@ -157,6 +162,67 @@ IPCReply GetRealProductCode(const IOCtlVRequest& request)
   return IPCReply(IPC_SUCCESS);
 }
 
+IPCReply SetDiscordClient(const IOCtlVRequest& request)
+{
+  if (!Config::Get(Config::MAIN_USE_DISCORD_PRESENCE))
+    return IPCReply(IPC_EACCES);
+
+  if (!request.HasNumberOfValidVectors(1, 0))
+    return IPCReply(IPC_EINVAL);
+
+  std::string new_client_id = Memory::GetString(request.in_vectors[0].address, request.in_vectors[0].size);
+
+  Discord::UpdateClientID(new_client_id);
+
+  return IPCReply(IPC_SUCCESS);
+}
+
+IPCReply SetDiscordPresence(const IOCtlVRequest& request)
+{
+  if (!Config::Get(Config::MAIN_USE_DISCORD_PRESENCE))
+    return IPCReply(IPC_EACCES);
+
+  if (!request.HasNumberOfValidVectors(10, 0))
+    return IPCReply(IPC_EINVAL);
+
+  std::string details = 
+      Memory::GetString(request.in_vectors[0].address, request.in_vectors[0].size);
```

Comment:
Do we want to set size limits on these?

The upstream API has size limits of 128 bytes on most fields (except the image keys which are 32 bytes).

Reply:
There's (seemingly) no negative side effect to sending an invalid presence update - the backend just denies it silently with no feedback to the guest (which is what this mimicks right now). I personally think the guest code should be responsible for complying with the limits at the time of development (which *may* change with updates, even if highly unlikely)

Reply:
I'm a little more worried about a corrupted ioctl potentially causing dolphin to crash, rather than the what the upstream library thinks. 

But it's probably not worth worrying about.